design-proposals/003-credentials.md

```diff
@@ -0,0 +1,226 @@
+
+The backend model is essentially a set of key-value pairs identified by the same **id** to be used in the view layer. The keys correspond to property names needed in connector configuration (i.e. set in `application.properties`) and the values are obtained during credential acquisition.
+
+As some secrets are not exposed to the UI, e.g. secrets that are private to Syndesis such as OAuth client secret or secret/private key, the same backend model constructed during the creation of kubernetes deployment descriptor (i.e. project generation).
+
+
+## Implementation
+
+The Credential provider should implement the following Java interface:
+
+```java
+interface CredentialProvider {
+    /**
+     * All AcquisitionMethod that this CredentialProvider supports
+     */
+    AcquisitionMethod[] acquisitionMethods();
+    /**
+     * Begin acquisition of the credential by returning the information
+     * to the UI on how to proceed and make any arrangements needed 
+     */
+    Acquisition acquire(String selectedMethod, Map<String, String> methodParameters, ContinuationStore continuationStore);
+    /** 
+     * Process any acquisition callbacks that might occur during the
+     * acquisition and produce Credentials.
+     */
+    Credential process(HttpServletRequest request, ContinuationStore continuationStore);
+    /*
+     * Gives any Credentials that are needed for deployment of the
+     * integration.
+     */
+    Credential[] credentialsFor(Integration integration);
+}
+```
+```java
+/**
+ * Information on the method of credential acquisition, needed for UI
+ * to present it to the user.
+ */
+class AcquisitionMethod {
+    String type;
+    String label;
+    String icon;
+    String description;
+}
+```
+```java
+interface Acquisition {}
+class DirectAcquisition extends Acquisition {
+    Credentials credentials; 
+}
+class CallbackAcquisition extends Acquisition {
+    String action;
+    String url;
+}
+```
+```java
+class Credential {
+    String id;
+    Map<String, String> content;
+}
+```
+
+## REST API
+
+To interact with the credentials API:
+
+| HTTP Verb | Path | Description |
+| --------- | ---- | ----------- |
+| GET | /api/{version}/connectors/{connector}/credentials | List all the ways a credential can be acquired for a {connector} |
+| POST | /api/{version}/connectors/{connector}/credentials | Acquire new credential for connector with label {connector} |
+| \* | /api/{version}/connectors/{connector}/credentials/\*\* | Handle interaction with 3rd party services during credential acquisition |
+| GET | /api/{version}/credentials | List all credentials created by the current tennant |
+| DELETE | /api/{version}/credentials/{id} | Remove credential from the system |
+
+### Examples
```

Comment:
A sequence diagram would be awesome. You can easily do it e.g. with http://plantuml.com/sequence-diagram